### PR TITLE
Update NN database with usage

### PR DESCRIPTION
**Pull request by tomtor**
Implements #758

### CODE REVIEW

**Comment by tomtor** [comment]
@vondele WIP, still have to show the data on the NN stats page....

File: fishtest/fishtest/views.py [hunk 1]
```diff
@@ -360,6 +362,24 @@ def get_sha(branch, repo_url):
     return '', ''
 
 
+def get_net(branch, repo_url):
+  """ Get the net from ucioption.cc in the repo """
```
Comment (vondele) [inline fishtest/fishtest/views.py:366]: ucioption.cc -> ucioption.cpp

File: fishtest/fishtest/views.py [hunk 2]
```diff
@@ -558,6 +578,22 @@ def del_tasks(run):
     del run['tasks']
   return run
 
+
+def update_nets(request, run_id, data):
+  if data['base_tag'] == 'master':
+    global MASTER_NET
+    MASTER_NET = get_net('master', data['tests_repo'])
```
Comment (vondele) [inline fishtest/fishtest/views.py:585]: Does this take care that this 'master' branch / commit is also part of the official stockfish, i.e. I could push a net to my master branch on a private repo and start a test, but that wouldn't make it official. The test approval page shows if base is in master or not, so much have some logic there.

File: fishtest/fishtest/views.py (hunk 1 above)
Comment (tomtor) [reply fishtest/fishtest/views.py:366]: fixed

File: fishtest/fishtest/views.py (hunk 2 above)
Comment (tomtor) [reply fishtest/fishtest/views.py:585]: `validate_form()` checks for equal signature, but also has ` data['base_same_as_master']` so I will use that instead.
Thanks for this feedback!

Comment (vondele) [reply fishtest/fishtest/views.py:585]: btw, I now see you use MASTER_NET as a global, i was rather thinking that any net that ever has been in the master repository would be marked in the database (i.e. db['is_master'] = db['is_master'] or data['base_same_as_master']). In that way we could mark all nets that have been 'default net' in the history of development differently in the list. The current default in master will probably always be the one with the most recent submission date.

Comment (tomtor) [reply fishtest/fishtest/views.py]: Ok, changed that...

**Comment by tomtor** [comment]
@ppigazzini Should be ready for some testing

**Comment by ppigazzini** [comment]
@tomtor DEV updated

**Comment by vondele** [comment]
looks good in my testing:

image

the two default nets get green, a tested  but non-default net remains white.

One question, do we want to add the feature that users that are not logged in do not see the 'untested' nets ? That would be the trick to make the illegal uploads less interesting.

**Comment by ppigazzini** [comment]
@vondele I started only two tests (SPRT, NumGames) with high priority, both with "master" vs "master", I'm wondering why are marked several net with the time stamp:
https://dfts-0.pigazzini.it/tests/view/5f36b1d01822588a3156711e
https://dfts-0.pigazzini.it/tests/view/5f36b1e91822588a31567120

**Comment by tomtor** [comment]
> One question, do we want to add the feature that users that are not logged in do not see the 'untested' nets ? That would be the trick to make the illegal uploads less interesting.

It would still be downloadable from the API URL.

I am not sure why these nets are uploaded (what does the uploader gain?)

Any way, it's easy to add if it makes sense.

**Comment by vondele** [comment]
I also submitted two tests (to get the older master and the tested non-master net)

Actually that raises another question. Do we updated the database when the test is submitted or when it is approved. I don't have too strong an opinion.

**Comment by tomtor** [comment]
I don't have to change any code when we update when submitted ;-)

**Comment by vondele** [comment]
> > One question, do we want to add the feature that users that are not logged in do not see the 'untested' nets ? That would be the trick to make the illegal uploads less interesting.
> 
> It would still be downloadable from the API URL.
> 
> I am not sure why these nets are uploaded (what does the uploader gain?)
> 
> Any way, it's easy to add if it makes sense.

yes, they could be downloaded if their sha is known, however, we don't 'advertise' or offer them for download. We're somewhat more on the safer side with content that is not clearly licensed. So, if easy, I would do it.

I'm also not sure what the uploader gains, but I think he enjoys the public visibility of the net, possibly uses it do promote distribution. .. No idea really.

**Comment by vondele** [comment]
> I don't have to change any code when we update when submitted ;-)

good enough a reason. At that point we anyway have the github handle of the person. It would become listed on the page before an approver can have a look, however.

**Comment by tomtor** [comment]
> It would become listed on the page before an approver can have a look, however.

I will change it...

**Comment by ppigazzini** [comment]
DEV updated.

**Comment by vondele** [comment]
and it works :-)

**Comment by tomtor** [comment]
Should update now on approval....

**Comment by tomtor** [comment]
Hmm. The exceptions (no net) don't make sense on approval.

Could move one to the new test code.

**Comment by vondele** [comment]
actually net == None can happen for tests running on older versions. However if net is not none, the net should be in the database. If net is not none, and the net could not be found in the database, it should not be possible to submit the test (I think).

**Comment by vondele** [comment]
One question I would have is how to add some meaningful data to the prod server once we apply this. I can, once merged on master, easily run a few tests on prod with old master so all former default nets get some 'green' label, and a few tests on a branch for the other net so they get the 'tested' status and become visible. Or did you have a different plan?

**Comment by tomtor** [comment]
> One question I would have is how to add some meaningful data to the prod server once we apply this. I can, once merged on prod, easily run a few tests on prod with old stockfish master so all former default nets get some 'green' label, and a few tests on a branch for the other net so they get the 'tested' status and become visible. Or did you have a different plan?

We could also write a small script to update the current NN database on prod, but I prefer approving some real tests (the script is a bit more work)

**Comment by vondele** [comment]
yes, that's what I was thinking as well.

**Comment by ppigazzini** [comment]
DEV updated, clean MongoDB synced with PROD.

**Comment by tomtor** [comment]
@ppigazzini @vondele For me it's OK to merge. Did some testing and updating stats on Approval looks OK.

**Comment by ppigazzini** [comment]
PROD updated, thank you @tomtor :)
@vondele please run the tests to populate the NN db on PROD :)

**Comment by vondele** [comment]
OK, will do now.

**Comment by tomtor** [comment]
No green labels on prod?

**Comment by vondele** [comment]
yes, to my surprise. Something seems not quite OK with the 'master' detection

**Comment by vondele** [comment]
the commits sha are from master, but of course are not HEAD on master. That's something we need to be able to deal with, if I do multiple commits in one push, a net might become default but never be 'publically' HEAD.

**Comment by tomtor** [comment]
@vondele You would have to help me with the github query. I could manually update NN DB to add the green flag to at least one net.

**Comment by vondele** [comment]
I'll try to help figure out, what do you want to do exactly?

**Comment by vondele** [comment]
(right now no need to manually make things green, let's see if we can fix it, and I can run the tests again).

**Comment by tomtor** [comment]
The reschedules of old tests on dev to mark nets green DID work.

**Comment by vondele** [comment]
Maybe related to how I created the tests on master? I didn't reschedule but used the sha (so it is not a branch name master).

**Comment by vondele** [comment]
actually, that's probably it, I have always used SF11 as base... stupid me.

**Comment by vondele** [comment]
I'll submit new tests, reverted.

**Comment by tomtor** [comment]
@vondele Looks good!

**Comment by vondele** [comment]
yes, now next step, I'll try to submit tests to get the tested nets visible, and not green :-)
this will take a little more time, need to do some searching in history.

**Comment by vondele** [comment]
so, seems to work. As far as I know the database is now complete. 
I think sergio didn't test nn-5a3983417e4f.nnue 
and nn-f6005244adb0.nnue Amplaytro is not tested either yet.

**Comment by vondele** [comment]
seemingly the test based on the wip-nnue-player branch didn't compile/work added a final test for nn-c157e0a5755b.nnue

Thanks for the work, guys!

**Pull request merged**